File: data_structures/BinaryTreeNode.ts
export interface BinaryTreeNode<T> {
  val: T;
  left: BinaryTreeNode<T> | null;
  right: BinaryTreeNode<T> | null;
}

type isLeft = boolean;

export function createTreeNode<T>(values: T[]): BinaryTreeNode<T> {
  const root: BinaryTreeNode<T> = createSingleNode(values[0])!;
  let queue: [BinaryTreeNode<T>, isLeft][] = [[root, true], [root, false]];

  for (let i = 1; i < values.length; i++) {
    const [node, isLeft] = queue.shift()!;
    const childNode = createSingleNode(values[i]);
    if (isLeft) {
      node.left = childNode;
    } else {
      node.right = childNode;
    }
    if (childNode) {
      queue.push([childNode, true], [childNode, false]);
    }
  }

  return root;
}

export const createSingleNode = <T>(value: T | null): BinaryTreeNode<T> | null => {
  return (value === null) ? null : { val: value, left: null, right: null };
};

File: data_structures/BinaryTreeNode.go
package datastructures

// BinaryTreeNode literally indicates a node of a binary tree.
type BinaryTreeNode struct {
	Val   int
	Left  *BinaryTreeNode
	Right *BinaryTreeNode
}

func createBinaryTreeNode(nums []int) BinaryTreeNode {
	if len(nums) == 0 {
		return BinaryTreeNode{0, nil, nil}
	}

	root := BinaryTreeNode{nums[0], nil, nil}

	queue := []*BinaryTreeNode{&root}

	for len(nums) > 0 {
		node, queue := queue[0], queue[1:]
		leftVal, rightVal, nums := nums[0], nums[1], nums[2:]

		if leftVal > 0 {
			node.Left = &BinaryTreeNode{leftVal, nil, nil}
			queue = append(queue, node.Left)
		}
		if rightVal > 0 {
			node.Right = &BinaryTreeNode{rightVal, nil, nil}
			queue = append(queue, node.Right)
		}
	}

}
